tests/e2e/migrate/public_operator_to_cockroach_enterprise_operator_test.go

```diff
@@ -0,0 +1,125 @@
+}
+
+func TestPublicOperatorToCockroachEnterpriseOperator(t *testing.T) {
+	h := newPublicOperatorToCockroachEnterpriseOperator()
+	h.TestDefaultMigration(t)
+}
+
+func (o *PublicOperatorToCockroachEnterpriseOperator) TestDefaultMigration(t *testing.T) {
+	o.Namespace = "cockroach" + strings.ToLower(random.UniqueId())
+	kubectlOptions := k8s.NewKubectlOptions("", "", o.Namespace)
+	k8s.CreateNamespace(t, k8s.NewKubectlOptions("", "", o.Namespace), o.Namespace)
+
+	// Create cluster with different logging config than the default one.
+	createLoggingConfig(t, k8sClient, "logging-config", o.Namespace)
+
+	crdbClusterName := "crdb-test"
+	o.CustomResourceBuilder = optestutil.NewBuilder(crdbClusterName).
+		WithNodeCount(3).
+		WithTLS().
+		WithImage(migration.CockroachVersion).
+		WithPVDataStore("1Gi").
+		WithLabels(map[string]string{"app": "cockroachdb"}).
+		WithAnnotations(map[string]string{"crdb": "isCool"}).
+		WithTerminationGracePeriodSeconds(int64(10)).
+		WithResources(corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("200m"),
+				corev1.ResourceMemory: resource.MustParse("256Mi"),
+			},
+		}).WithClusterLogging("logging-config")
+
+	o.Ctx = context.Background()
+	o.CrdbCluster = testutil.CockroachCluster{
+		Cfg:              cfg,
+		K8sClient:        k8sClient,
+		StatefulSetName:  crdbClusterName,
+		Namespace:        o.Namespace,
+		ClientSecret:     fmt.Sprintf("%s-root", crdbClusterName),
+		NodeSecret:       fmt.Sprintf("%s-node", crdbClusterName),
+		CaSecret:         fmt.Sprintf("%s-ca", crdbClusterName),
+		IsCaUserProvided: true,
+		Context:          k3dClusterName,
+		DesiredNodes:     3,
+	}
+
+	o.InstallOperator(t)
+
+	testutil.RequireClusterToBeReadyEventuallyTimeout(t, o.CrdbCluster, 600*time.Second)
+	time.Sleep(20 * time.Second)
+	testutil.RequireCRDBToFunction(t, o.CrdbCluster, false)
+	testutil.RequireCRDBDatabaseToFunction(t, o.CrdbCluster, migration.TestDBName, "")
+
+	prepareForMigration(t, o.CrdbCluster.StatefulSetName, o.Namespace, o.CrdbCluster.CaSecret, "operator")
+	defer func() {
+		_ = os.RemoveAll(manifestsDirPath)
+	}()
+
+	o.UninstallOperator(t)
+	t.Log("Create the priority class crdb-critical which will be owned by the cockroachdb enterprise operator")
+	k8s.RunKubectl(t, kubectlOptions, "create", "priorityclass", "crdb-critical", "--value", "500000000")
+	defer k8s.RunKubectl(t, kubectlOptions, "delete", "priorityclass", "crdb-critical")
+
+	t.Log("Create the rbac permissions for the cockroachdb enterprise operator")
+	k8s.KubectlApply(t, kubectlOptions, filepath.Join(manifestsDirPath, "rbac.yaml"))
+
+	t.Log("Install the cockroachdb enterprise operator")
+	operator.InstallCockroachDBEnterpriseOperator(t, kubectlOptions)
+	defer func() {
+		t.Log("Uninstall the cockroachdb enterprise operator")
+		operator.UninstallCockroachDBEnterpriseOperator(t, kubectlOptions)
+	}()
+
+	migratePodsToCrdbNodes(t, o.CrdbCluster.StatefulSetName, o.Namespace)
```

Comment:
I'm wondering if we can make more assertions here. For example, should we assert that all crdb pods are ready after each one is updated? Assert that we don't see underreplicated ranges during the migration? Insert some data pre-migration and assert it's still present afterwards? One easy change might be to run `o.ValidateCRDB(t)` after each node is updated.

Reply:
Yes, that is the plan. However this PR in itself became quite huge. So I am going to add more assertions in follow up PRs. I can do ValidateCRDB post every replica update in this PR.